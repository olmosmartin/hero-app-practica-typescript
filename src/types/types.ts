export const types = {
    login: 'SET_AUTH',
    logout: 'CLEAR_AUTH',
};