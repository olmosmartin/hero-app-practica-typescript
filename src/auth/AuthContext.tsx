import { createContext } from 'react'
import { AuthContextType } from '../models/AuthContextType';

const UserState = {logged:false, user:""}

export const AuthContext = createContext<AuthContextType>({user:UserState, dispatch:()=>{}});
