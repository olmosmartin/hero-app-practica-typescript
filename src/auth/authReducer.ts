import { types } from "../types/types";
import { ActionAuthReducer } from "../models/ActionAuthReducer";

/*
const state = {
    usuario: "",
    logged: false
}
*/


export const authReducer = (state = {}, action:ActionAuthReducer) => {

    switch (action.type) {
        case types.login:
            return {
                ...state,
                ...action.payload,
                logged: true
            };
        case types.logout:
            return {
                logged: false
            };
        default:
            return state;
    }

}