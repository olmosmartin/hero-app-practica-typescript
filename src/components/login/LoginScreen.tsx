import { useHistory } from "react-router-dom";



export const LoginScreen = () => {
    const history = useHistory()
    
    const handleOnClick = () => {
        // replace es para que no quede en el stack de navegacion?
        history.replace('/');
    }

    return (
        <div className='container mt-5'>
            <h1>LoginScreen</h1>
            <hr />

            <button
                className='btn btn-primary'
                onClick={handleOnClick}
            >
                Login
            </button>
        </div>
    )
}
