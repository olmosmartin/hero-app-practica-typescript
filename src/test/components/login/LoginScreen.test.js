import { mount } from "enzyme";
import '@testing-library/jest-dom';
import { LoginScreen } from "../../../components/login/LoginScreen";
import { AuthContext } from "../../../auth/AuthContext";
import { types } from "../../../types/types";


describe('pruebas en LoginScreen.tsx', () => {

    const history = {
        replace: jest.fn(),
    }

    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged: false,
            user: "Martin"
        }
    }

    const wrapper = mount(
        <AuthContext.Provider value={ contextValue } >
            <LoginScreen history = { history } />
        </AuthContext.Provider>
    );


    test('debe mostrarse correctamente con el snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('debe realizar el dispatch y la navegacion', () => {

        wrapper.find('button').prop('onClick')();
        
        expect(contextValue.dispatch).toHaveBeenCalledWith({
            type: types.login,
            payload: {
                user: "Martin"
            }
        })
        expect(wrapper).toMatchSnapshot();
        expect( history.replace ).toHaveBeenCalled();

    })

    test('debe guardar el local storage', () => {

        const handleClick = wrapper.find('button').prop('onClick');
        
        handleClick();

        expect(contextValue.dispatch).toHaveBeenCalledWith({
            type: types.login,
            payload: {
                user: "Martin"
            }
        })
        expect( history.replace ).toHaveBeenCalledWith("/");

        localStorage.setItem('lastPath', '/dc')
        handleClick();

        expect( history.replace ).toHaveBeenCalledWith("/dc");

    })

})