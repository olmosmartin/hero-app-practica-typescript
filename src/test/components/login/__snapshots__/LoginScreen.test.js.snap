// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en LoginScreen.tsx debe mostrarse correctamente con el snapshot 1`] = `
<div
  className="container mt-5"
>
  <h1>
    LoginScreen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;

exports[`pruebas en LoginScreen.tsx debe realizar el dispatch y la navegacion 1`] = `
<div
  className="container mt-5"
>
  <h1>
    LoginScreen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
